File: benchmarks/R/run-bakeoff.R
#!/usr/bin/env Rscript
# R-side bake-off runner.
#
# Usage:
#   Rscript benchmarks/R/run-bakeoff.R --out OUT.csv [--only id1,id2,...] [--fixture NAME] [--par]
#
# Reads benchmarks/queries.yaml, opens each required fixture once (unless
# reopen=true in the entry, in which case the fixture is reopened inside
# the benched expression), executes each query via bench::mark, emits a
# CSV aligned with the Julia runner.

options(error = NULL)  # user's .Rprofile may set recover()

suppressPackageStartupMessages({
    library(dafr)
    library(bench)
    library(yaml)
    library(digest)
})

args <- commandArgs(trailingOnly = TRUE)
.arg <- function(flag, default = NULL) {
    i <- match(flag, args)
    if (is.na(i) || i == length(args)) return(default)
    args[i + 1L]
}
.has_flag <- function(flag) flag %in% args

out_path    <- .arg("--out", "/tmp/r-times.csv")
only_ids    <- .arg("--only", NULL)
only_fixt   <- .arg("--fixture", NULL)
min_iter    <- as.integer(.arg("--min-iter", "5"))
par_mode    <- .has_flag("--par")

# ---- single-threaded baseline unless --par ----
if (!par_mode) {
    Sys.setenv(OMP_NUM_THREADS = "1")
    options(dafr.kernel_threshold = Inf)   # force scalar code paths
    if (requireNamespace("RhpcBLASctl", quietly = TRUE)) {
        RhpcBLASctl::blas_set_num_threads(1)
        RhpcBLASctl::omp_set_num_threads(1)
    }
}

fixture_root <- "benchmarks/fixture/data"
queries <- yaml::read_yaml("benchmarks/queries.yaml")

if (!is.null(only_ids)) {
    ids <- strsplit(only_ids, ",", fixed = TRUE)[[1L]]
    queries <- Filter(function(q) q$id %in% ids, queries)
}
if (!is.null(only_fixt)) {
    queries <- Filter(function(q) q$fixture == only_fixt, queries)
}
cat(sprintf("running %d queries\n", length(queries)))

.open_fixture <- function(name) {
    path <- file.path(fixture_root, name)
    if (name == "chain_triple" || name == "view_renamed") {
        dafr::complete_daf(file.path(path, "leaf"), "r")
    } else {
        dafr::files_daf(path, name = name, mode = "r")
    }
}

fixtures_needed <- unique(vapply(queries, function(q) q$fixture, character(1L)))
opened <- list()
for (name in fixtures_needed) {
    opened[[name]] <- .open_fixture(name)
    cat(sprintf("opened fixture %s\n", name))
}

# ---- fixture checksums (sha256 over sorted file content) ----
.sha256_dir <- function(path) {
    files <- sort(list.files(path, recursive = TRUE, full.names = TRUE))
    hashes <- vapply(files, function(f) digest::digest(file = f, algo = "sha256"),
                     character(1L))
    digest::digest(paste(basename(files), hashes, collapse = "\n"), algo = "sha256", serialize = FALSE)
}
checksums <- vapply(sort(fixtures_needed),
                    function(n) substr(.sha256_dir(file.path(fixture_root, n)), 1, 16),
                    character(1L))

# ---- benchmark loop ----
rows <- vector("list", length(queries))
for (k in seq_along(queries)) {
    q <- queries[[k]]
    reopen <- isTRUE(q$reopen)
    text <- q$text
    fixture_name <- q$fixture

    # Build the benched expression. For reopen=true, re-open the fixture
    # inside the benched expression so I/O is part of what we measure.
    if (reopen) {
        expr <- bquote({
            d <- .open_fixture(.(fixture_name))
            dafr::get_query(d, .(text))
        })
    } else {
        d <- opened[[fixture_name]]
        expr <- bquote(dafr::get_query(.(d), .(text)))
    }

    t0 <- Sys.time()
    b <- tryCatch(
        bench::mark(
            eval(expr),
            min_iterations = min_iter,
            filter_gc = FALSE,
            check = FALSE,
            time_unit = "ns"
        ),
        error = function(e) {
            cat(sprintf("  [%3d/%3d] %-30s %-22s FAIL: %s\n",
                        k, length(queries), q$id, q$category,
                        conditionMessage(e)))
            NULL
        }
    )
    if (is.null(b)) {
        rows[[k]] <- data.frame(
            query_id = q$id, query_text = q$text,
            category = q$category, fixture = q$fixture,
            median_time_ns = NA_real_, min_time_ns = NA_real_,
            gc_count = NA_integer_, allocations = NA_real_,
            n_iter = NA_integer_,
            stringsAsFactors = FALSE
        )
        next
    }
    dt <- as.numeric(Sys.time() - t0, units = "secs")
    cat(sprintf("  [%3d/%3d] %-30s %-22s ok (%.1fs)\n",
                k, length(queries), q$id, q$category, dt))

    rows[[k]] <- data.frame(
        query_id       = q$id,
        query_text     = q$text,
        category       = q$category,
        fixture        = q$fixture,
        median_time_ns = as.numeric(b$median),
        min_time_ns    = as.numeric(b$min),
        gc_count       = b$n_gc,
        allocations    = sum(as.numeric(b$mem_alloc), na.rm = TRUE),
        n_iter         = b$n_itr,
        stringsAsFactors = FALSE
    )
}

df <- do.call(rbind, rows)

# ---- header lines ----
commit <- tryCatch(
    system("git rev-parse HEAD", intern = TRUE)[1L],
    error = function(e) "unknown"
)
header_lines <- c(
    sprintf("# runner: R"),
    sprintf("# dafr_commit: %s", commit),
    sprintf("# R_version: %s", getRversion()),
    sprintf("# platform: %s", R.version$platform),
    sprintf("# OMP_NUM_THREADS: %s", Sys.getenv("OMP_NUM_THREADS", "default")),
    sprintf("# BLAS: %s", extSoftVersion()[["BLAS"]]),
    sprintf("# fixtures: %s", paste(sprintf("%s=%s", names(checksums), checksums), collapse = "; "))
)

dir.create(dirname(out_path), recursive = TRUE, showWarnings = FALSE)
writeLines(header_lines, out_path)
suppressWarnings(
    write.table(df, out_path, sep = ",", row.names = FALSE, col.names = TRUE,
                append = TRUE, qmethod = "double")
)
cat(sprintf("\nwrote %s (%d rows)\n", out_path, nrow(df)))

File: benchmarks/julia/run_bakeoff.jl
#!/usr/bin/env julia
# Julia-side bake-off runner.
#
# Usage:
#   julia --project=benchmarks/julia benchmarks/julia/run_bakeoff.jl --out OUT.csv
#       [--only id1,id2,...] [--fixture NAME] [--par]

using DataAxesFormats
using BenchmarkTools
using YAML
using CSV
using DataFrames
using SHA
using LinearAlgebra
using ArgParse
using Printf

function parse_args_()
    s = ArgParseSettings()
    @add_arg_table! s begin
        "--out";     arg_type = String; default = "/tmp/julia-times.csv"
        "--only";    arg_type = String; default = ""
        "--fixture"; arg_type = String; default = ""
        "--par";     action = :store_true
    end
    return parse_args(s)
end

const ARGS_ = parse_args_()

if !ARGS_["par"]
    BLAS.set_num_threads(1)
    # JULIA_NUM_THREADS is set at launch, not runtime. Just record what we got.
end

const FIXTURE_ROOT = joinpath(@__DIR__, "..", "fixture", "data")

function sha256_dir(path::AbstractString)
    # Mirror R's digest-based approach exactly:
    #   files <- sort(list.files(path, recursive=TRUE, full.names=TRUE))
    #   hashes <- vapply(files, function(f) digest::digest(file=f, algo="sha256"), character(1))
    #   digest::digest(paste(basename(files), hashes, collapse="\n"), algo="sha256")
    # paste(v1, v2, collapse="\n") produces "b1 h1\nb2 h2\n..." (space-sep pairs, \n between, no trailing \n).
    files = String[]
    for (root, _, fs) in walkdir(path)
        for f in fs
            push!(files, joinpath(root, f))
        end
    end
    sort!(files)
    pairs = String[]
    for f in files
        fhash = open(f) do io
            bytes2hex(SHA.sha256(io))
        end
        push!(pairs, "$(basename(f)) $fhash")
    end
    combined = join(pairs, "\n")
    return bytes2hex(SHA.sha256(Vector{UInt8}(combined)))[1:16]
end

function open_fixture(name::AbstractString)
    path = joinpath(FIXTURE_ROOT, name)
    if name == "chain_triple" || name == "view_renamed"
        return complete_daf(joinpath(path, "leaf"), "r"; name = name)
    else
        return FilesDaf(path, "r"; name = name)
    end
end

function main()
    queries_all = YAML.load_file(joinpath(@__DIR__, "..", "queries.yaml"))
    queries = queries_all
    if !isempty(ARGS_["only"])
        ids = Set(split(ARGS_["only"], ","))
        queries = filter(q -> q["id"] in ids, queries)
    end
    if !isempty(ARGS_["fixture"])
        queries = filter(q -> q["fixture"] == ARGS_["fixture"], queries)
    end
    println("running $(length(queries)) queries")

    fixtures_needed = sort(unique(q["fixture"] for q in queries))
    opened = Dict{String,Any}()
    for name in fixtures_needed
        opened[name] = open_fixture(name)
        println("opened fixture $name")
    end
    checksums = Dict(name => sha256_dir(joinpath(FIXTURE_ROOT, name))
                     for name in fixtures_needed)

    rows = DataFrame(query_id = String[], query_text = String[],
                     category = String[], fixture = String[],
                     median_time_ns = Float64[], min_time_ns = Float64[],
                     gc_count = Int[], allocations = Float64[],
                     n_iter = Int[])

    for (k, q) in enumerate(queries)
        text      = q["text"]
        name      = q["fixture"]
        reopen    = get(q, "reopen", false)

        bench = try
            if reopen
                @benchmark get_query(open_fixture($name), $text) samples=50 seconds=5 evals=1
            else
                daf = opened[name]
                @benchmark get_query($daf, $text) samples=50 seconds=5 evals=1
            end
        catch e
            @printf "  [%3d/%3d] %-30s %-22s FAIL: %s\n" k length(queries) q["id"] q["category"] sprint(showerror, e)
            push!(rows, (q["id"], text, q["category"], name,
                         NaN, NaN, 0, NaN, 0))
            continue
        end

        push!(rows, (q["id"], text, q["category"], name,
                     Float64(median(bench).time),
                     Float64(minimum(bench).time),
                     Int(length(filter(x -> x > 0, bench.gctimes))),
                     Float64(bench.allocs),
                     length(bench.times)))

        @printf "  [%3d/%3d] %-30s %-22s ok\n" k length(queries) q["id"] q["category"]
    end

    commit = try
        strip(read(`git rev-parse HEAD`, String))
    catch
        "unknown"
    end

    # Sorted fixtures line for join-match with R runner
    fixture_line = join(["$k=$(checksums[k])" for k in sort(collect(keys(checksums)))], "; ")

    blas_info = try
        string(BLAS.get_config())
    catch
        "unknown"
    end

    header = [
        "# runner: Julia",
        "# dafr_commit: $commit",
        "# julia_version: $(VERSION)",
        "# platform: $(Sys.MACHINE)",
        "# JULIA_NUM_THREADS: $(get(ENV, "JULIA_NUM_THREADS", "default"))",
        "# BLAS: $blas_info",
        "# fixtures: $fixture_line",
    ]

    mkpath(dirname(ARGS_["out"]))
    open(ARGS_["out"], "w") do io
        for line in header
            println(io, line)
        end
        CSV.write(io, rows, append = true, writeheader = true)
    end
    println("\nwrote $(ARGS_["out"]) ($(nrow(rows)) rows)")
end

main()
